### PR TITLE
Meson multi-precision

### PR DESCRIPTION
**Pull request by amontoison**
### Status

- [ ] amd
- [ ] ampl
- [ ] apps
- [ ] arc
- [ ] bgo
- [ ] blls
- [ ] bqp
- [ ] bqpb
- [ ] bsc
- [ ] buildspec
- [ ] ccqp
- [ ] cdqp
- [ ] check
- [ ] checkpoint
- [ ] clls
- [x] clock
- [x] common
- [ ] convert
- [ ] copyright
- [ ] cqp
- [ ] cro
- [ ] cutest_functions
- [ ] demo
- [ ] dgo
- [ ] dlp
- [ ] dps
- [ ] dqp
- [ ] dum
- [ ] eqp
- [ ] extend
- [ ] external
- [ ] fdc
- [ ] fdh
- [ ] filter
- [ ] filtrane
- [ ] fit
- [ ] glrt
- [ ] gls
- [ ] gltr
- [ ] hash
- [ ] hsl
- [ ] ir
- [ ] l1qp
- [ ] l2rt
- [ ] lancelot
- [ ] lancelot_simple
- [x] lapack
- [ ] lhs
- [ ] lls
- [ ] llst
- [ ] lms
- [ ] lmt
- [ ] lpa
- [ ] lpb
- [ ] lpqp
- [ ] lqr
- [ ] lqt
- [ ] lsp
- [ ] lsqp
- [ ] lsrb
- [ ] lsrt
- [ ] lstr
- [ ] matlab
- [ ] miqr
- [ ] mop
- [ ] nlpt
- [ ] nls
- [ ] norms
- [ ] opt
- [x] precision
- [ ] presolve
- [ ] problem
- [ ] psls
- [ ] ptrans
- [ ] qp
- [ ] qpa
- [ ] qpb
- [ ] qpc
- [ ] qpd
- [ ] qpp
- [ ] qpt
- [ ] rand
- [ ] roots
- [ ] rpd
- [ ] rqs
- [ ] sbls
- [ ] scale
- [ ] scaling
- [ ] scu
- [ ] sec
- [ ] sha
- [ ] sils
- [ ] slls
- [ ] sls
- [x] smt
- [ ] sort
- [x] space
- [x] specfile
- [ ] spral
- [ ] ssids
- [ ] string
- [x] symbols
- [ ] tools
- [ ] trans
- [ ] trb
- [ ] trs
- [ ] tru
- [x] ugo
- [ ] uls
- [x] userdata
- [ ] warm
- [ ] wcp
- [x] zd11

### CODE REVIEW

**Comment by dpo** [comment]
Why annihilate all the work I've spent many days on? Just submit your pr to the `meson-build` branch.

**Comment by amontoison** [comment]
@nimgould updates all packages to support single and double precision. It breaks GALAHAD, so I created a new branch to stay synchronize with the modifications of Nick.
I reuse what you did on the `meson-build`, I just add them one by one on this branch.

File: src/ugo/ugo.F90 [hunk 1]
```diff
@@ -2189,7 +2189,7 @@ SUBROUTINE UGO_solve_direct( data, userdata, status, x, f, g, h, eval_FGH )
 
      INTERFACE
        SUBROUTINE eval_FGH( status, x, userdata, f, g, h )
-       USE GALAHAD_USERDATA_double
+       USE GALAHAD_USERDATA_precision
```
Comment (amontoison) [inline src/ugo/ugo.F90:2192]: Nick, this is the culprit of our issue of Friday! A `GALAHAD_USERDATA_double` was still hard coded in `ugo.F90`.
I suppose that you compile the double precision `.mod` before the single precision one on our computer.

**Comment by amontoison** [comment]
@nimgould 
Can you test this Meson script on your computer?
```
# Git
git checkout master
git fetch origin
git checkout -b meson-mp
git rebase origin/meson-mp

# Meson
CC=gcc FC=gfortran meson setup builddir
meson compile -C builddir
meson test -C builddir
```
The tests for the Fortran routines and the C interfaces passed in both single and double precision!
I just have some random errors with `ugot` in double precision, sometimes the test fails:
```
4/4 ugo_double          OK              0.03s

4/4 ugo_double          FAIL            0.14s   exit status 2
>>> MALLOC_PERTURB_=255 /home/alexis/Bureau/git/GALAHAD/builddir/ugot_double

4/4 ugo_double          FAIL            0.14s   exit status 2
>>> MALLOC_PERTURB_=244 /home/alexis/Bureau/git/GALAHAD/builddir/ugot_double

4/4 ugo_double          FAIL            0.15s   exit status 2
>>> MALLOC_PERTURB_=149 /home/alexis/Bureau/git/GALAHAD/builddir/ugot_double

4/4 ugo_double          OK              0.03s
```
Even if the tests passed in single precision, I had the following note when I ran `./ugot_single` in the folder `builddir` directly:
`Note: The following floating-point exceptions are signalling: IEEE_UNDERFLOW_FLAG IEEE_DENORMAL`.

**Comment by dpo** [comment]
> @nimgould updates all packages to support single and double precision. It breaks GALAHAD, so I created a new branch to stay synchronize with the modifications of Nick.
> I reuse what you did on the `meson-build`, I just add them one by one on this branch.
> Two shared libraries are generated here and it's only working for UGO.

Yes I fully appreciate that. However I spent many days understanding Meson and writing a first build script. It would be polite to preserve my commits in the commit history :-).

**Comment by amontoison** [comment]
Dominique, this pull request has just created to easily visualize the modified Meson script for the branch `multi-precision` and our small proof of concept with UGO. The diff is with the branch `multi-precision` such Nick and Jari can easily read it.

The goal is not to appropriate your work...
I sincerely hope that it's not what you think otherwise I won't be happy.
We are just prototyping for now.

Please wait that I send you a report to fully understand what we decided with Nick and Jari this week and what we plan to do.

**Comment by nimgould** [comment]
After I checked out the meson-mp branch, I get an error

% git checkout -b meson-mp
Switched to a new branch 'meson-mp'
% git rebase origin/meson-mp
fatal: invalid upstream 'origin/meson-mp'

Any ideas why?

**Comment by nimgould** [comment]
Dominique, sorry for the silence from this end, Covid intervened for a few days.

The week with Alexis was most productive. The major decision is to move any dependence of the GALAHAD packages on sed to build precision-dependent versions from a prototype, but rather to use (c) preprocessing directives to do so for us; although these aren't standard in fortran, every sensible compiler we know of does. Currently we allow for a variety of real precisions (currently single and double, but easily extendable to half and quadruple if they are supported) and integer word lengths (currently 4 and 8 bytes, but again extendable). I'm slowly working through all 600 or so GALAHAD source files making the changes (using sed to help!!), but this shouldn't take more than a couple of weeks, excluding Christmas. We are very keen to switch to meson, and although we may have broken things at present, it is clear that this is the way forward. I'm sure that there are many challenges on the way!

**Comment by amontoison** [comment]
> After I checked out the meson-mp branch, I get an error
> 
> % git checkout -b meson-mp Switched to a new branch 'meson-mp' % git rebase origin/meson-mp fatal: invalid upstream 'origin/meson-mp'
> 
> Any ideas why?

I think that you forgot the command `git fetch origin`.

File: src/ugo/ugo.F90 (hunk 1 above)
Comment (nimgould) [reply src/ugo/ugo.F90:2192]: Indeed, poor old Harry (he plays for my club, and usually scores good penalties!). I would hate to take a penalty against my own team mate (Hugo Lloris also plays for us)

**Comment by nimgould** [comment]
That works, thank you.

I get 

Ok:                 3   
Expected Fail:      0   
Fail:               1   

and the failure seems to be line 936 of file ../src/ugo/ugo.F90 (unit = 40)
Fortran runtime error: Cannot open file 'ALIVE.d'. This is very odd as the line 
before checks to see if the file exists, and only opens it if it doesn't. Is this
test being done in parallel? Can I just run the last (single precision) test on its own?

**Comment by amontoison** [comment]
Yes, the tests are run in parallel. You can just run the last single precision test on its own, the shared libraries and tests are binaries are in the  `$GALAHAD/builddir` folder.

**Comment by nimgould** [comment]
Thanks again. As I suspected, when I just run this test (it was the double precision version, notg the single one, my mistake), all is well. This is an issue as fortran has no way of knowing if a file has been added/deleted by another program at the same time, unless it is run in its own workspace. Is it possible to move to a separate folder for each test?

**Comment by amontoison** [comment]
I didn't find (yet) how to create a workspace for each test.
However I found how to run tests in sequential and create test suites:
```
test('ugo_single', ugot_single, suite: 'single', is_parallel : false)
```
With `is_parallel : false`, Meson will not run another test in the same time.
With the `suite` argument, we can also run all single or double precision tests with the following command:
```
meson test -C builddir --suite single
```
And if we just want to run one specific test, we can use:
```
meson test -C builddir ugo_single
```

**Comment by nimgould** [comment]
Great. Since we are only interested in whether the tests work (and they shouldn't take any time at all), we should run these (but not the codes themselves) sequentially. Thanks for finding this out, @amontoison

**Comment by amontoison** [comment]
Nick, I have some issues with the Julia interface of UGO, I have the same segmentation fault in single and double precision.
`ugo_initialize at ..../src/ugo/C/ugo_ciface.F90:295`

I think that you modified some fields in the C structures. I remarked that `second_derivative_available` was removed  from `ugo_control_type` in `ugo_ciface.F90` but we still have it in `include/galahad_ugo.h`, which I use to generate the Julia wrappers with Clang.jl.

**Comment by amontoison** [comment]
superseded by #26

**Pull request closed without merging**